filters/include/pcl/filters/experimental/grid_filter.h

```diff
@@ -0,0 +1,234 @@
+   * with  \a setFilterLimits, points having values outside this interval will be
+   * discarded. \param[in] field_name the name of the field that contains values used
+   * for filtering
+   */
+  inline void
+  setFilterFieldName(const std::string& field_name)
+  {
+    filter_field_name_ = field_name;
+  }
+
+  /** \brief Get the name of the field used for filtering. */
+  inline std::string const
+  getFilterFieldName() const
+  {
+    return filter_field_name_;
+  }
+
+  /** \brief Get the field filter limits (min/max) set by the user. The default values
+   * are -FLT_MAX, FLT_MAX. \param[out] limit_min the minimum allowed field value
+   * \param[out] limit_max the maximum allowed field value
+   */
+  inline void
+  getFilterLimits(double& limit_min, double& limit_max) const
+  {
+    limit_min = filter_limit_min_;
+    limit_max = filter_limit_max_;
+  }
+
+  /** \brief Set to true if we want to return the data outside the interval specified by
+   * setFilterLimits (min, max). Default: false. \param[in] limit_negative return data
+   * inside the interval (false) or outside (true)
+   */
+  inline void
+  setFilterLimitsNegative(const bool limit_negative)
+  {
+    filter_limit_negative_ = limit_negative;
+  }
+
+  /** \brief Get whether the data outside the interval (min/max) is to be returned
+   * (true) or inside (false). \param[out] limit_negative true if data \b outside the
+   * interval [min; max] is to be returned, false otherwise
+   */
+  inline void
+  getFilterLimitsNegative(bool& limit_negative) const
+  {
+    limit_negative = filter_limit_negative_;
+  }
+
+  /** \brief Get whether the data outside the interval (min/max) is to be returned
+   * (true) or inside (false). \return true if data \b outside the interval [min; max]
+   * is to be returned, false otherwise
+   */
+  inline bool
+  getFilterLimitsNegative() const
+  {
+    return filter_limit_negative_;
+  }
+
+protected:
+  /** \brief Set to true if all fields need to be downsampled, or false if just XYZ. */
+  bool downsample_all_data_;
+
+  /** \brief Set to true if leaf layout information needs to be saved in \a
+   * leaf_layout_. */
+  bool save_leaf_layout_;
+
+  /** \brief The desired user filter field name. */
+  std::string filter_field_name_;
+
+  /** \brief The minimum allowed filter value a point will be considered from. */
+  double filter_limit_min_;
+
+  /** \brief The maximum allowed filter value a point will be considered from. */
+  double filter_limit_max_;
+
+  /** \brief Set to true if we want to return the data outside (\a filter_limit_min_;\a
+   * filter_limit_max_). Default: false. */
+  bool filter_limit_negative_;
+
+  /** \brief Minimum number of points per voxel for the centroid to be computed */
+  unsigned int min_points_per_voxel_;
+
+  /** \brief The iterable grid object for storing information of each fraction of space
+   * in the filtering space defined by the grid */
+  Grid grid_;
```

Comment:
Should we move this to the GridStruct, and instead require a data member or a function to access that?

My line of thought:
* grid in user's class makes sense from code-dev perspective
* No need to pass it in all the functions

Reply:
This would also resolve the issue in https://github.com/PointCloudLibrary/pcl/pull/4813#discussion_r659007997

Reply:
Done, after moving it we can also eliminate a template argument of `GridFilter`